odl/operator/operator.py

```diff
@@ -95,57 +80,279 @@ def _default_apply(self, x, out, *args, **kwargs):
+    spec_msg += '\n'.join(py2_specs)
+    if py3:
+        spec_msg += '\n' + '\n'.join(py3only_specs)
+    spec_msg += '\n\nStatic or class methods are not allowed.'
+
+    if sum(arg is not None for arg in (cls, bound_call, unbound_call)) != 1:
+        raise ValueError('Exactly one object to check must be given.')
+
+    if cls is not None:
+        # Get the actual implementation, including ancestors
+        for parent in cls.mro():
+            call = parent.__dict__.get(attr, None)
+            if call is not None:
+                break
+
+        # Static and class methods are not allowed
+        if isinstance(call, staticmethod):
+            raise TypeError("'{}.{}' is a static method. "
+                            "".format(cls.__name__, attr) + spec_msg)
+        elif isinstance(call, classmethod):
+            raise TypeError("'{}.{}' is a class method. "
+                            "".format(cls.__name__, attr) + spec_msg)
+
+    elif bound_call is not None:
+        call = bound_call
+        if not inspect.ismethod(call):
+            raise TypeError('{} is not a bound method.'.format(call))
+    else:
+        call = unbound_call
+
+    # pylint: disable=deprecated-method,redefined-variable-type
+    if py3:
+        # support kw-only args and annotations
+        spec = inspect.getfullargspec(call)
+        kw_only = spec.kwonlyargs
+        kw_only_defaults = spec.kwonlydefaults
+    else:
+        spec = inspect.getargspec(call)
+        kw_only = ()
+        kw_only_defaults = {}
 
-    def __call__(cls, *args, **kwargs):
-        """Create a new class ``cls`` from given arguments.
+    signature = _signature_from_spec(call)
 
-        Raises
-        ------
-        NotImplementedError
-            If neither ``_call`` or ``_apply`` is implemented.
-        """
-        obj = type.__call__(cls, *args, **kwargs)
-        if not hasattr(obj, '_call') and not hasattr(obj, '_apply'):
-            raise NotImplementedError('`Operator` instances must'
-                                      'either have `_call` or `_apply`'
-                                      ' as attribute.')
-        return obj
+    pos_args = spec.args
+    if unbound_call is not None:
+        # Add 'self' to positional arg list to satisfy the checker
+        pos_args.insert(0, 'self')
+
+    pos_defaults = spec.defaults
+    varargs = spec.varargs
+
+    out_optional = False
+
+    # Variable args are not allowed
+    if varargs is not None:
+        raise ValueError("Bad signature '{}': variable arguments not allowed."
+                         " ".format(signature) + spec_msg)
 
+    if len(pos_args) not in (2, 3):
+        raise ValueError("Bad signature '{}'. ".format(signature) + spec_msg)
+
+    true_pos_args = pos_args[1:]
+    if len(true_pos_args) == 1:  # 'out' kw-only
+        if 'out' in true_pos_args:  # 'out' positional and 'x' kw-only -> no
+            raise ValueError("Bad signature '{}': `out` cannot be the only "
```

Comment:
See the comment. This is in the case where `true_pos_args` has length 1 (remember `pos_args[0] == 'self'`), so `out` is the only positional arg here, which we disallow.